docs/implementation/query-prefetching.md

```diff
@@ -0,0 +1,505 @@
+`store/postgres/src/relational_queries.rs`
+
+
+<a id="org60a4097"></a>
+
+## Handling first/skip
+
+We never get all the `children` for a parent; instead we always have a
+`first` and `skip` argument in the children filter. Those arguments need to
+be applied to each parent individually by ranking the children for each
+parent according to the order defined by the `children` query. If the same
+child matches multiple parents, we need to make sure that it is considered
+separately for each parent as it might appear at different ranks for
+different parents. In SQL, we use the `rank()` window function for this:
+
+    select *
+      from (
+        select c.*,
+               rank() over (partition by parent_id order by ...) as pos
+          from (query to get children) c)
+     where pos >= skip and pos < skip + first
+
+
+<a id="org926c9ee"></a>
+
+## Handling interfaces
+
+If `parents` or `children` (or both) are interfaces, we resolve the
+interfaces into the concrete types implementing them, produce a query for
+each combination of parent/child concrete type and combine those queries
+via `union all`.
+
+Since implementations of the same interface will generally differ in the schema
+they use, we can not form a `union all` of all the data in the tables for
+these concrete types, but have to first query only attributes that we know
+will be common to all entities implementing the interface, most notably the
+`vid` (a unique identifier that identifies the precise version of an
+entity), and then later fill in the details of each entity by converting it
+directly to JSON.
+
+That means that when we deal with children that are an interface, we will
+first select only the following columns (where exactly they come from
+depends on how the parent/child relationship is modeled)
+
+    select '{__typename}' as entity, c.vid, c.id, parent_id
+
+and form the `union all` of these queries. We then use that union to rank
+children as described above.
+
+
+<a id="org205427b"></a>
+
+## Handling parent/child relationships
+
+How we get the children for a set of parents depends on how the
+relationship between the two is modeled. The interesting parameters there
+are whether parents store a list or a single child, and whether that field
+is derived, together with the same for children.
+
+There are a total of 16 combinations of these four boolean variables; four
+of them, when both parent and child derive their fields, are not
+permissible. It also doesn't matter whether the child derives its parent
+field: when the parent field is not derived, we need to use that since that
+is the only place that contains the parent -> child relationship. When the
+parent field is derived, the child field can not be a derived field.
+
+That leaves us with the following combinations of whether the parent and
+child store a list or a scalar value, and whether the parent is derived:
+
+<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">
```

Comment:
Neat! I didn't know that is possible. Since we now have an RFC repo, I wonder if that doc should really go there - otherwise, it's going to be the only document about implementation in `graph-node`.

Reply:
It's not quite an engineering plan but I agree, it's a good place to put it. Could you create an engineering plan PR with this document as `PLAN-0001: SQL query combination`?